### PR TITLE
Improve ReadGate and SolutionReadout performance

### PR DESCRIPTION
**Pull request by AlessandroPierro**
Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
- [x] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`pyb`) passes locally
- [x] Build tests (`pyb -E unit`) or (`python -m unittest`) passes locally

## Pull request type

 

 

Please check your PR type:
- [x] Bugfix

## What is the new behavior?

- The handshake between ReadGate and SolutionReadout is removed, improving performance.
- ReadGate and SolutionReadout are updated to enable a new CProcModel for the ReadGate

## Does this introduce a breaking change?

- [x] No

### CODE REVIEW

File: src/lava/lib/optimization/solvers/generic/read_gate/process.py [hunk 1]
```diff
@@ -45,6 +45,8 @@ def __init__(self,
                          name=name,
                          log_config=log_config)
         self.target_cost = Var(shape=(1,), init=target_cost)
+        self.best_solution = Var(shape=shape, init=-1)
+        self.best_solution_step = Var(shape=(1,), init=-1)
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/read_gate/process.py:49]: Not sure if these need to be actual Vars (or LavaTypes in the ProcModel). IMHO, we can just define them as nparrays in the PyProcModel and not mention them in the Process. Vars are only necessary if you want to somehow interact with them from externally, like by defining their initial value or reading them during runtime.

Or am I missing something, @srrisbud ?

**Comment by phstratmann** [comment]
Good work!

I think you also need to push the changed CProcModels simultaneously. Otherwise, your code change here may break unit tests in Lava Framework. Or am I wrong?

**Comment by AlessandroPierro** [comment]
> Good work!
> 
> I think you also need to push the changed CProcModels simultaneously. Otherwise, your code change here may break unit tests in Lava Framework. Or am I wrong?

Thanks!

Yeah, totally right. I'm preparing that PR as well.

File: src/lava/lib/optimization/solvers/generic/read_gate/process.py (hunk 1 above)
Comment (AlessandroPierro) [reply src/lava/lib/optimization/solvers/generic/read_gate/process.py:49]: These are used in the CProcModel to store the best solution (and associated time step) across different iterations. Am I missing something or is it right to add them to the Process?

**Review by phstratmann — APPROVED** [review]

**Pull request merged**